nupic/research/frameworks/pytorch/imagenet/experiment_utils.py

```diff
@@ -345,3 +349,161 @@ def create_model(model_class, model_args, init_batch_norm, device,
             model = hook(model, **kwargs) or model
 
     return model
+
+
+def get_free_port():
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
+        # bind on port 0 - kernel will select an unused port
+        s.bind(("", 0))
+        # removed socket.SO_REUSEADDR arg
+        # TCP error due to two process with same rank in same port - maybe a fix
+        return s.getsockname()[1]
+
+
+class SearchOption:
+    """
+    Allow usage of random and grid search for exploration of hyperparameters space
+    In-place substitute for Ray Tune simlar options while not available
+    """
+
+    def __init__(self, elements):
+        self.elements = elements
+
+
+class RandomSearch(SearchOption):
+    """Sample from a list of options or a stochastic function"""
+
+    def expand_in_place(self, config, k1, k2=None):
+        # get element
+        if callable(self.elements):
+            element = self.elements()
+        elif hasattr(self.elements, "__iter__"):
+            element = np.random.choice(self.elements)
+        else:
+            raise ValueError(
+                "RandomSearch requires a function with no args or an iterator"
+            )
+
+        # assign
+        if k2 is None:
+            config[k1] = element
+        else:
+            config[k1][k2] = element
+
+
+class GridSearch(SearchOption):
+    """Expand one experiment per option"""
+
+    def expand_to_list(self, config, k1, k2=None):
+        expanded_list = []
+        for el in self.elements:
+            expanded_config = deepcopy(config)
+            if k2 is None:
+                expanded_config[k1] = el
+            else:
+                expanded_config[k1][k2] = el
+            expanded_list.append(expanded_config)
+
+        return expanded_list
+
+
+class TrialsCollection:
+    def __init__(self, config, num_trials, restore=True):
```

Comment:
Some docstring would be useful. What is this class supposed to do, and what do the parameters mean?

Reply:
Added docstring